### PR TITLE
GH-81: Infer output format based on file extension

### PR DESCRIPTION
**Pull request by adelhult**
Resolves GH-81.

### CODE REVIEW

**Comment by CMDJojo** [comment]
Should we infer output extension from format flag, if the output doesn't have an extension? Like should `modmark -f html in.mdm out` make a file `out.html`?

File: cli/src/main.rs [hunk 1]
```diff
@@ -44,10 +39,30 @@ struct Args {
     dev: bool,
 }
 
+// Infer the output format based on the file extension of the output format
+fn infer_output_format(output: &Path) -> Option<OutputFormat> {
+    output
+        .extension()
+        .map(|ext| match ext.to_str() {
+            Some("tex") => Some(OutputFormat::new("latex")),
+            Some("html") => Some(OutputFormat::new("html")),
```
Comment (CMDJojo) [inline cli/src/main.rs:48]: Maybe add htm as well?

File: cli/src/main.rs [hunk 2]
```diff
@@ -44,10 +39,30 @@ struct Args {
     dev: bool,
 }
 
+// Infer the output format based on the file extension of the output format
+fn infer_output_format(output: &Path) -> Option<OutputFormat> {
+    output
+        .extension()
+        .map(|ext| match ext.to_str() {
```
Comment (CMDJojo) [inline cli/src/main.rs:46]: `and_then` to not have to map

File: cli/src/main.rs [hunk 3]
```diff
@@ -44,10 +39,30 @@ struct Args {
...
+}
+
 fn compile_file(args: &Args) -> Result<Ast, CliError> {
     let source = fs::read_to_string(&args.input)?;
     let mut ctx = Context::default();
-    let output = eval(&source, &mut ctx, &OutputFormat::new(&args.format))?;
+
+    let Some(format) = args.format
+        .as_ref()
+        .map(|s| OutputFormat::new(&s))
```
Comment (CMDJojo) [inline cli/src/main.rs:60]: Hmm, you do `as_ref` and then take a ref, seems to me that you can replace `&s` to `s`, and then u are fine with just giving the function handle directly

**Review by CMDJojo — CHANGES_REQUESTED** [review]
Seems nice, good work! I found some code style tings I think should be fixed before merging, as well as my previous comment which would be nice to implement, if you agree that that is a good idea

File: cli/src/main.rs (hunk 3 above)
Comment (adelhult) [reply cli/src/main.rs:60]: Nope, sadly it doesn't work since `OutputFormat::new` expects a `&str` and I have `&String`.

Comment (CMDJojo) [reply cli/src/main.rs:60]: It is the same thing with automatic coersion, and `.map(|s| OutputFormat::new(s))` compiles on my machine

**Review by CMDJojo — APPROVED** [review]
Looks good now! (For other's information, we talked in private and concluded that while document programs often automatically add file extensions, command line tools doesn't so often, so we skipped it here for now). Ready to merge!

**Pull request merged**